mirgecom/steppers.py

```diff
@@ -59,6 +58,17 @@ def advance_state(rhs, timestepper, checkpoint, get_timestep,
         Time at which to start
     istep: int
         Step number from which to start
+    pre_step_callback
+        An optional user-defined function to be called before the timestepper
+        is called for that particular step. A non-zero return code from this
+        function indicates that this function should stop gracefully.
+        Examples of such functions include visualization, io, or restart
+        checkpoints.
+    post_step_callback
+        An optional user-defined function to be called after the timestepper
+        is called for that particular step. A non-zero return code from this
+        function indicates that this function should stop gracefully.
+        Examples of such functions include applying modal filtering or limiters.
```

Comment:
Modifying state would entail returning it. It's also a dangerous thing to do in the setting of multi-step time steppers which maintain their own history.

Reply:
Agreed, and we need to take care with that.  But some of the things we've done in the (near and distant) past involve just that: actuators or filters that modify the state between steps.

These user hooks could be a nice way for the users to test different sorts of filters or controls in optimization problems.  As @inducer notes, we'd need to allow the callback to modify the state and return it and that's not in the signature (yet).  And folks would need to be mindful about what sorts of timestepping would make sense.

Reply:
@inducer How are we supposed to apply limiters then? These modify the solution. I don't see a way around that. I agree we should be careful though.

Reply:
I'm just grumpy that we need a facility like this. I'm not arguing that we shouldn't have one. :) The one substantive bit to my comment is that the callback should return the modified solution, since in-place modification is not a thing.

Reply:
https://github.com/illinois-ceesd/mirgecom/pull/370/commits/b914d3dd258983b3bbf07a5e315a85579b5565bc